src/components/tooltip/tooltip.js

```diff
@@ -3,391 +3,428 @@
-        var rawText = override || element.text().trim();
-        var interpolatedText = $interpolate(rawText)(parent.scope());
-        parent.attr('aria-label', interpolatedText);
-      }
-    }
+          // To avoid 'synthetic clicks', we listen to mousedown instead of
+          // 'click'.
+          parent.on('mousedown', mousedownEventHandler);
+          parent.on(ENTER_EVENTS, enterEventHandler);
 
-    function manipulateElement () {
-      element.detach();
-      element.attr('role', 'tooltip');
-    }
+          function windowScrollEventHandler() {
+            setVisible(false);
+          }
 
-    function bindEvents () {
-      var mouseActive = false;
-
-      // add an mutationObserver when there is support for it
-      // and the need for it in the form of viable host(parent[0])
-      if (parent[0] && 'MutationObserver' in $window) {
-        // use an mutationObserver to tackle #2602
-        var attributeObserver = new MutationObserver(function(mutations) {
-          if (mutations.some(function (mutation) {
-              return (mutation.attributeName === 'disabled' && parent[0].disabled);
-            })) {
-              $mdUtil.nextTick(function() {
-                setVisible(false);
-              });
+          function windowBlurEventHandler() {
+            elementFocusedOnWindowBlur = document.activeElement === parent[0];
           }
-        });
 
-        attributeObserver.observe(parent[0], { attributes: true});
-      }
+          function enterEventHandler($event) {
+            // Prevent the tooltip from showing when the window is receiving
+            // focus.
+            if ($event.type === 'focus' && elementFocusedOnWindowBlur) {
+              elementFocusedOnWindowBlur = false;
+            } else if (!scope.mdVisible) {
+              parent.on(LEAVE_EVENTS, leaveEventHandler);
+              setVisible(true);
+
+              // If the user is on a touch device, we should bind the tap away
+              // after the 'touched' in order to prevent the tooltip being
+              // removed immediately.
+              if ($event.type === 'touchstart') {
+                parent.one('touchend', function() {
+                  $mdUtil.nextTick(function() {
+                    $document.one('touchend', leaveEventHandler);
+                  }, false);
+                });
+              }
+            }
+          }
 
-      // Store whether the element was focused when the window loses focus.
-      var windowBlurHandler = function() {
-        elementFocusedOnWindowBlur = document.activeElement === parent[0];
-      };
+          function leaveEventHandler() {
+            autohide = scope.hasOwnProperty('mdAutohide') ?
+                scope.mdAutohide :
+                attr.hasOwnProperty('mdAutohide');
+
+            if (
+              autohide ||
+              mouseActive ||
+              $document[0].activeElement !== parent[0]
```

Comment:
Helpful advice via Clang format (I ran it on the entire file - here's the result for this method):

```
      function leaveEventHandler() {
        autohide = scope.hasOwnProperty('mdAutohide') ?
            scope.mdAutohide :
            attr.hasOwnProperty('mdAutohide');

        if (autohide || mouseActive ||
            $document[0].activeElement !== parent[0]) {
          // When a show timeout is currently in progress, then we have
          // to cancel it, otherwise the tooltip will remain showing
          // without focus or hover.
          if (showTimeout) {
            $timeout.cancel(showTimeout);
            setVisible.queued = false;
            showTimeout = null;
          }

          parent.off(LEAVE_EVENTS, leaveEventHandler);
          parent.triggerHandler('blur');
          setVisible(false);
        }
        mouseActive = false;
      }
```